src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp

```diff
@@ -273,74 +320,171 @@ void G1BarrierSetAssembler::g1_write_barrier_post(MacroAssembler* masm,
   assert(thread == r15_thread, "must be");
 #endif // _LP64
 
-  Address queue_index(thread, in_bytes(G1ThreadLocalData::dirty_card_queue_index_offset()));
-  Address buffer(thread, in_bytes(G1ThreadLocalData::dirty_card_queue_buffer_offset()));
+  Label done, runtime;
 
-  CardTableBarrierSet* ct =
-    barrier_set_cast<CardTableBarrierSet>(BarrierSet::barrier_set());
+  Assembler::Condition is_single_region = generate_single_region_test(masm, store_addr, new_val, tmp);
+  __ jcc(is_single_region, done);
 
-  Label done;
+  Assembler::Condition is_new_val_null = generate_new_val_null_test(masm, new_val);
```

Comment:
I actually kind of like the previous style that (almost) all "asm" is at the same level -- the newly introduced helper functions hinder the flow, IMO.

Reply:
Note that if we want to optimize the barrier code layout (see the [JEP description](https://openjdk.org/jeps/475), *Candidate optimizations* sub-section), splitting the assembly of each barrier in at least two blocks is necessary, since we need to separate the inline from the out-of-line (barrier stub) code. And since the assembly code has to be split into multiple functions anyway, I think it makes sense to group the code by logical blocks (different barrier tests, queue insertion, etc.), as proposed in this changeset. This also improves code reuse, e.g. the same `generate_queue_insertion` implementation is used for the pre- and post-barriers.
If you still think there is value in grouping together the blocks that can be grouped together (e.g. `generate_single_region_test` + `generate_new_val_null_test` + `generate_card_young_test`), I can prototype the refactoring and let the G1 maintainers decide which alternative is more readable/maintainable.

Reply:
> This also improves code reuse

In this area, I think code duplication is less of an issue -- it's more crucial that one can follow the asm flow as if reading real asm. (Ofc, this is subjective; feel free to keep as is.)

Reply:
I personally find the current style more maintainable in that it 1) makes the high-level structure of the barriers more explicit and 2) reuses code across the pre- and post-barrier (the fact that queue insertion is an identical operation for both barriers is obscured in mainline by naming and instruction choice differences). Having said that, if there is consensus among G1 maintainers that generating the fast (inline) and slow (out of line) paths of each barrier in single functions (sacrificing reuse of the queue insertion logic) is preferable, I'm happy to change that. @tschatzl @kimbarrett what do you think?

Reply:
I'm back from vacation now and resuming my work in this JEP. After some offline discussions, I have pushed a new version (commit 20ef68c81e) without helper functions, except for `generate_queue_insertion()` which is still included. @albertnetymk please have a look and let me know if you find the new style more readable.